Improves handling of Gcode Thumbnails

pedrolamas opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

This PR adds improved handling of Gcode Thumbnails that include JPG and QOI image formats ~, and also adds Gcode files support to Peek~.

![image](https://github.com/microsoft/PowerToys/assets/85504/aa3a9f2c-c23c-4b48-bbe9-a27aa0e57d80)

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #28404
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [X] **Tests:** Added/updated and all pass
- [X] **Localization:** All end user facing strings can be localized
- [X] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

PowerToys already had support to show PNG thumbnails found in Gcode files as File Explorer thumbnails or on the preview pane.

As Prusa Slicer recently added JPG and QOI image formats to the output Gcode, this PR adds support for those 2 new image formats.

The QOI image format support is added via QoiImage, a custom QOI decoder I wrote based on the [reference code](https://github.com/phoboslab/qoi/blob/master/qoi.h).

I've refactored the Gcode parsing related code and moved it to PreviewHandlerCommon.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Manual testing on my local machine mostly, but I also updated the `UnitTests-GcodePreviewHandler` and `UnitTests-GcodeThumbnailProvider` tests

pedrolamas committed:
Improves handling of Gcode Thumbnails

pedrolamas commented:
I would argue that the Peek support doesn't bring that much value as these thumbnails are normally 128x128 or less, so we end up with a small image shown on Peek!

**However**, g-code files have a lot more metadata than just the thumbnails, so if peek had a Property Grid of sorts, we could show that metadata as key value pairs... just a thought for further enhancement! 😉

pedrolamas committed:
Remove Peek support

pedrolamas commented:
Following up on my previous comment, I see I was incorrectly adding Peek support via ImagePreviewer in this is no image, so I have reverted and removed those changes from the PR.

pedrolamas commented:
I've just added #27979 to allow Peek to show not only Gcode thumbnails, but also any file that has a thumbnail!

pedrolamas committed:
Merge branch 'main' into pedrolamas/gcode-thumbnails-extra

pedrolamas committed:
Moves GcodeHelper to PreviewHandlerCommon

jaimecbernardo reviewed (changes requested):
Thanks a lot for the contribution!
Blocking until .74 while we figure out how to audit the new dependency ;)

stefansjfw commented:
> > Blocking until .74 while we figure out how to audit the new dependency ;)
> 
> @jaimecbernardo if the WTFPL license is an issue, there are a couple of other NuGet packages that are MIT based, maybe I should check those instead?
> 
> Alternatively, I can try pulling a QOI implementation into PowerToys...

What alternate packages do you have in ming?

stefansjfw commented:
> Alternatively, I can try pulling a QOI implementation into PowerToys

How much code would this bring in? If not much, this might be the best option. The reason is that all of these packages have a small number of downloads, and not sure how stable are those

jaimecbernardo commented:
With the WTFPL license, I'm thinking this could be a git submodule, so we can be sure of the code we are bringing in.
My biggest issue here is that it's low number of downloads so we might be bringing some risky code in that would be signed and running on end user machines.

pedrolamas commented:
I will take a look at one of the other MIT licensed QOI implementations and see if I can get my code to work with theirs; if so, I expect it to be one or two files tops to bring into PT, so might that be workable?

pedrolamas committed:
Reverts minor change

jaimecbernardo commented:
That would be an interesting way of doing it, as well. I'm planning on making the current dependency into a git submodule after I do the current round of reviews :)

pedrolamas committed:
Merge remote-tracking branch 'upstream/main' into pedrolamas/gcode-thumbnails-extra

pedrolamas committed:
Skip unknown data on GcodeHelper.GetBestThumbnail

pedrolamas committed:
Replaces QOI.Core with QoiImage

pedrolamas committed:
Merge remote-tracking branch 'upstream/main' into pedrolamas/gcode-thumbnails-extra

pedrolamas commented:
FWIW, I've now removed the QOI.Core dependency and replaced it with my very own custom implementation of a QOI decoder based on the (MIT licensed) reference code!

pedrolamas committed:
Fixes spellchecker

pedrolamas committed:
Reverts changes to NOTICE.md

pedrolamas committed:
Minor QoiImage improvements

pedrolamas committed:
Use custom QoiPixel struct

jaimecbernardo commented:
Hi @pedrolamas , thanks for the changes :) 
So this is based on https://github.com/phoboslab/qoi , right?

pedrolamas commented:
> Hi @pedrolamas , thanks for the changes :) So this is based on [phoboslab/qoi](https://github.com/phoboslab/qoi) , right?

Correct, this is my own translation of the official reference code:

![image](https://github.com/microsoft/PowerToys/assets/85504/90207f91-5921-4e59-9a50-22866a30b5e3)

I added a small disclaimer on top of the QoiImage.cs file, pointing to the reference code file.

jaimecbernardo committed:
Add MIT notice for the QOI reference code

jaimecbernardo commented:
Thanks a lot for the implementation 😄 
I've added the MIT notice to our NOTICE file.
@crutkas , I believe this should be enough to comply with the MIT license for the reference code that was translated and adapted?
https://github.com/microsoft/PowerToys/pull/27947/commits/3c9a04796973fac4320a38f26fd201487c82499a

jaimecbernardo committed:
Fix spellcheck for the MIT notice

crutkas commented:
Can you link me to their license md?

pedrolamas commented:
@crutkas https://github.com/phoboslab/qoi/blob/master/LICENSE

crutkas commented:
Lemme do the notice tweak and will call out it is a port too by Pedro

jaimecbernardo commented:
> Lemme do the notice tweak and will call out it is a port too by Pedro

FYI, This is what's already on the PR in NOTICE.md under "File Explorer add-ons":
```
### The Quite OK Image Format reference decoder

[@pedrolamas](https://github.com/pedrolamas) translated and adapted the reference decoder C++ code to C#.

**Source**: https://github.com/phoboslab/qoi

MIT License

Copyright (c) 2022 Dominic Szablewski

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

jaimecbernardo reviewed (approved):
LGTM! Thank you for the contribution!
Just waiting on @crutkas OK for the Notice.md changes and then we can merge 😉

crutkas committed:
Update NOTICE.md

tweaked notice a bit

crutkas commented:
i tweaked it a bit, @pedrolamas did great and on inspection, it would have been truthful.
